examples/tutorials/core_api_pytorch_mnist/model_def_distributed.py

```diff
@@ -0,0 +1,278 @@
+
+    test_loss /= len(test_loader.dataset)
+
+    print(
+        "\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n".format(
+            test_loss, correct, len(test_loader.dataset), 100.0 * correct / len(test_loader.dataset)
+        )
+    )
+
+    # NEW: Report metrics only on rank 0.
+    if core_context.distributed.rank == 0:
+        core_context.train.report_validation_metrics(
+            steps_completed=steps_completed, metrics={"test_loss": test_loss}
+        )
+
+    return test_loss
+
+
+def load_state(checkpoint_directory):
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath("checkpoint.pt").open("rb") as f:
+        return torch.load(f)
+
+
+def main(core_context):
+    # Training settings
+    parser = argparse.ArgumentParser(description="PyTorch MNIST Example")
+    parser.add_argument(
+        "--batch-size",
+        type=int,
+        default=64,
+        metavar="N",
+        help="input batch size for training (default: 64)",
+    )
+    parser.add_argument(
+        "--test-batch-size",
+        type=int,
+        default=1000,
+        metavar="N",
+        help="input batch size for testing (default: 1000)",
+    )
+    parser.add_argument(
+        "--epochs",
+        type=int,
+        default=14,
+        metavar="N",
+        help="number of epochs to train (default: 14)",
+    )
+    parser.add_argument(
+        "--lr", type=float, default=1.0, metavar="LR", help="learning rate (default: 1.0)"
+    )
+    parser.add_argument(
+        "--gamma",
+        type=float,
+        default=0.7,
+        metavar="M",
+        help="Learning rate step gamma (default: 0.7)",
+    )
+    parser.add_argument(
+        "--no-cuda", action="store_true", default=False, help="disables CUDA training"
+    )
+    parser.add_argument(
+        "--no-mps", action="store_true", default=True, help="disables macOS GPU training"
+    )
+    parser.add_argument(
+        "--dry-run", action="store_true", default=False, help="quickly check a single pass"
+    )
+    parser.add_argument("--seed", type=int, default=1, metavar="S", help="random seed (default: 1)")
+    parser.add_argument(
+        "--log-interval",
+        type=int,
+        default=100,
+        metavar="N",
+        help="how many batches to wait before logging training status",
+    )
+
+    args = parser.parse_args()
+    use_cuda = not args.no_cuda and torch.cuda.is_available()
+    use_mps = not args.no_mps and torch.backends.mps.is_available()
+
+    info = det.get_cluster_info()
+    assert info is not None, "this example only runs on-cluster"
+    latest_checkpoint = info.latest_checkpoint
+    if latest_checkpoint is not None:
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            model = load_state(path)
+
+    torch.manual_seed(args.seed)
+
+    if use_cuda:
+        device = torch.device("cuda")
```

Comment:
Can you check something for me?  Can you check if this is actually putting your model on each device?  My understanding is you're going to put all replicas on `cuda:0` implicitly with this code.  Since it's MNIST, you won't notice from running out of memory or anything, but I'm not sure this is being distributed to multiple nodes correctly.

Reply:
Interestingly enough it was putting all replicas on `cuda:0`, but it didn't matter because each GPU was on a different node anyway. Changed it to be `core_context.distributed.local_rank `which is more generally correct.